algebra/vector_space.lean

```diff
@@ -0,0 +1,305 @@
+import algebra.ring
+import algebra.module
+
+universes u' v'
+
+class vector_space (α : inout Type u')  (β : Type v') [field α] extends module α β
+
+instance field.to_vector_space {K : Type} [field K] :
+vector_space K K := ⟨ring.to_module⟩
+
+lemma eq_zero_of_add_self_eq {α : Type} [add_group α]
+{a : α} (H : a + a = a) : a = 0 :=
+add_left_cancel (by {rw add_zero, exact H})
+
+class subspace (K : Type) (V : Type) [field K] [vector_space K V] (p : V → Prop) :=
+    (p_zero : p 0)
+    (p_add : ∀ u v, p u → p v → p (u + v))
+    (p_neg : ∀ v, p v → p (-v))
+    (p_smul : ∀ c v, p v → p (c • v))
+
+namespace subspace
+
+variables (K : Type) {V : Type} [field K] [vector_space K V]
+{p : V → Prop} [subspace K V p]
+
+variables {c d : K} (u v : {x // p x})
+
+include K
+
+def add (u v : {x // p x}) : {x // p x} :=
+{
+    val := u.val + v.val,
+    property := p_add K u.val v.val u.property v.property
+}
+
+def zero : {x // p x} :=
+{
+    val := 0,
+    property := p_zero K p
+}
+
+def neg (v : {x // p x}) : {x // p x} :=
+{
+    val := -v.val,
+    property := p_neg K v.val v.property
+}
+
+def smul (c : K) (v : {x // p x}) : {x // p x} :=
+{
+    val := c • v.val,
+    property := p_smul c v.val v.property
+}
+
+instance : has_add {x // p x} := ⟨add K⟩
+instance : has_zero {x // p x} := ⟨zero K⟩
+instance : has_neg {x // p x} := ⟨neg K⟩
+instance : has_scalar K {x // p x} := ⟨smul K⟩
+
+@[simp] lemma add_simp : (add K u v).val = u.val + v.val := rfl
+@[simp] lemma zero_simp : (zero K : {x // p x}).val = 0 := rfl
+@[simp] lemma neg_simp : (neg K v).val = -v.val := rfl
+@[simp] lemma smul_simp : (smul K c v).val = c • v.val := rfl
+@[simp] lemma add_simp' : (u + v).val = u.val + v.val := rfl
+@[simp] lemma zero_simp' : (0 : {x // p x}).val = 0 := rfl
+@[simp] lemma neg_simp' : (-v).val = -v.val := rfl
+@[simp] lemma smul_simp' : (c • v).val = c • v.val := rfl
+
+instance : vector_space K {x // p x} :=
+{
+    add                 := add K,
+    add_assoc           := (λ u v w, subtype.eq (by simp [add_assoc])),
+    zero                := zero K,
+    zero_add            := (λ v, subtype.eq (by simp [zero_add])),
+    add_zero            := (λ v, subtype.eq (by simp [add_zero])),
+    neg                 := neg K,
+    add_left_neg        := (λ v, subtype.eq (by simp [add_left_neg])),
+    add_comm            := (λ u v, subtype.eq (by simp [add_comm])),
+    smul                := smul K,
+    smul_left_distrib   := (λ c u v, subtype.eq (by simp [smul_left_distrib])),
+    smul_right_distrib  := (λ c u v, subtype.eq (by simp [smul_right_distrib])),
+    mul_smul            := (λ c d v, subtype.eq (by simp [mul_smul])),
+    one_smul            := (λ v, subtype.eq (by simp [one_smul]))
+}
+
+end subspace
+
+
+structure linear_space (K V W : Type) [field K]
+[vector_space K V] [vector_space K W] :=
+    (T : V → W)
+    (map_add : ∀ u v, T (u+v) = T u + T v)
+    (map_mul : ∀ (c:K) v, T (c • v) = c • (T v))
```

Comment:
* Style: indent the second line (with the binders), but not the others
* There should be a `coe_to_fun` instance so that one can refer to `A x` instead of `A.T x`